src/main/java/de/tum/in/www1/artemis/service/programming/AuxiliaryRepositoryService.java

```diff
@@ -0,0 +1,209 @@
+            List<AuxiliaryRepository> updatedAuxiliaryRepositories, ProgrammingExercise updatedExercise) {
+        // Get all repositories that are unchanged and are still present in the updated exercise
+        List<AuxiliaryRepository> auxiliaryRepositories = new ArrayList<>(
+                Objects.requireNonNullElse(programmingExercise.getAuxiliaryRepositories(), new ArrayList<AuxiliaryRepository>()).stream()
+                        .filter(existingRepo -> updatedAuxiliaryRepositories.stream().noneMatch((updatedRepo -> existingRepo.getId().equals(updatedRepo.getId())))
+                                && updatedExercise.getAuxiliaryRepositories().stream().anyMatch(updatedRepo -> existingRepo.getId().equals(updatedRepo.getId())))
+                        .toList());
+
+        for (AuxiliaryRepository repo : updatedAuxiliaryRepositories) {
+            validateAuxiliaryRepository(repo, auxiliaryRepositories,
+                    programmingExercise.getAuxiliaryRepositories().stream().noneMatch(existingRepo -> existingRepo.getId().equals(repo.getId())));
+            auxiliaryRepositories.add(repo);
+        }
+        updatedExercise.setAuxiliaryRepositories(new ArrayList<>());
+        auxiliaryRepositories.forEach(updatedExercise::addAuxiliaryRepository);
+    }
+
+    private void validateAuxiliaryRepositoryId(AuxiliaryRepository auxiliaryRepository) {
+        if (auxiliaryRepository.getId() != null) {
+            throw new BadRequestAlertException("Auxiliary repositories must not have an id.", AUX_REPO_ENTITY_NAME,
+                    ProgrammingExerciseResource.ErrorKeys.INVALID_AUXILIARY_REPOSITORY_ID);
+        }
+    }
+
+    private void validateAuxiliaryRepositoryNameExists(AuxiliaryRepository auxiliaryRepository) {
+        if (auxiliaryRepository.getName() == null || auxiliaryRepository.getName().isEmpty()) {
+            throw new BadRequestAlertException("Cannot set empty name for auxiliary repositories!", AUX_REPO_ENTITY_NAME,
+                    ProgrammingExerciseResource.ErrorKeys.INVALID_AUXILIARY_REPOSITORY_NAME);
+        }
+        auxiliaryRepository.setName(auxiliaryRepository.getName().toLowerCase());
+    }
+
+    private void validateAuxiliaryRepositoryNameLength(AuxiliaryRepository auxiliaryRepository) {
+        if (auxiliaryRepository.getName().length() > AuxiliaryRepository.MAX_NAME_LENGTH) {
+            throw new BadRequestAlertException("The name of an auxiliary repository must not be longer than 100 characters!", AUX_REPO_ENTITY_NAME,
+                    ProgrammingExerciseResource.ErrorKeys.INVALID_AUXILIARY_REPOSITORY_NAME);
+        }
+    }
+
+    private void validateAuxiliaryRepositoryNameDuplication(AuxiliaryRepository auxiliaryRepository, List<AuxiliaryRepository> otherRepositories) {
+        for (AuxiliaryRepository existingRepository : otherRepositories) {
+            if (existingRepository.getName().equals(auxiliaryRepository.getName())) {
```

Comment:
Very minor nitpick: `auxiliaryRepository.getName()` could be extracted into a variable like in the method below for consistency reasons. Maybe it also would make sense for `validateAuxiliaryRepositoryNameExists()`.